Reformat the way optimization algorithms are defined (#592)
deepinv/deepinv: docs/source/user_guide/reconstruction/unfolded.rst

Unfolded Algorithms
===================

> The :ref:`optimization module <optim>` module also be turned into an unfolded trainable architecture.

Reviewer: ```suggestion
```
I would maybe remove this line since it is explained below?

Unfolded architectures (sometimes called 'unrolled architectures') are obtained by replacing parts of these algorithms
by learnable modules. In turn, they can be trained in an end-to-end fashion to solve inverse problems.
It suffices to set the argument ``unfold=True`` when creating an optimization algorithm from the :ref:`optimization <optim>` module.

> By default, if a neural network is used in place of a regularization or data-fidelity step, the parameters of the network are learnable by default.

Reviewer: ```suggestion
By default, if a neural network is used in place of a regularization or data-fidelity step, the parameters of the network are learnable.
```

> Moreover, among all the parameters of the algorithm (e.g. step size, regularization parameter, etc.), the use can use which are learnable and which are not via the argument ``trainable_params``.

Reviewer: ```suggestion
Moreover, among all the parameters of the algorithm (e.g. step size, regularization parameter, etc.), the user can choose which are learnable and which are not via the argument ``trainable_params``.
```

and :math:`x^\star` is the equilibrium point of the forward pass.
See `this tutorial <http://implicit-layers-tutorial.org/deep_equilibrium_models/>`_ for more details.

For turning an optimization algorithm into a DEQ model, the ``DEQ`` argument of :class:`deepinv.optim.BaseOptim` must be an instance of :class:`deepinv.optim.DEQConfig`, which defines the parameters for equilibrium-based implicit differentiation.
The :class:`deepinv.optim.DEQConfig` dataclass has the following attributes and default values:

.. code-block:: python

    @dataclass
    class DEQConfig:
        jacobian_free: bool = False

>             # Whether to use a Jacobian-free backward pass (see :footcite:t:`fung2022jfb`).

Reviewer: ```suggestion
            # Whether to use a Jacobian-free backward pass.
```
citations don't seem to work inside the code-block

            # Number of past iterates used in Anderson acceleration for the forward pass.
        beta_anderson_acc_forward: float = 1.0
            # Momentum coefficient in Anderson acceleration for the forward pass.
        eps_anderson_acc_forward: float = 1e-4
            # Regularization parameter for Anderson acceleration in the forward pass.

        # Backward pass Anderson acceleration
        anderson_acceleration_backward: bool = False
            # Whether to use Anderson acceleration for solving the backward equilibrium.
        history_size_backward: int = 5
            # Number of past iterates used in Anderson acceleration for the backward pass.
        beta_anderson_acc_backward: float = 1.0
            # Momentum coefficient in Anderson acceleration for the backward pass.
        eps_anderson_acc_backward: float = 1e-4
            # Regularization parameter for Anderson acceleration in the backward pass.
        max_iter_backward: int = 50
            # Maximum number of iterations in the backward equilibrium solver.

By default, DEQ is disabled (``DEQ=None``). As soon as ``DEQ`` is not ``None``, the above ``DEQConfig`` values are used.

> For now DEQ is only possible with Gradient Descent :class:`deepinv.optim.GD`,  Proximal Gradient Descent :class:`deepinv.optim.PGD` and Half-Quadratic-Splitting :class:`deepinv.optim.HQS` optimization algorithms.

Reviewer: ```suggestion
.. note::
             Currently, DEQ is only possible with :class:`Gradient Descent <deepinv.optim.GD>`,  :class:`Proximal Gradient Descent <deepinv.optim.PGD>` and :class:`Half-Quadratic-Splitting <deepinv.optim.HQS>` optimization algorithms.
```
